### PR TITLE
[GFX-1868] Fix W3 level MSVC warnings in public headers

### PR DESCRIPTION
**Pull request by palver123**
## Jira ticket URL (Why?)
[GFX-1868](https://shapr3d.atlassian.net/browse/GFX-1868)

## Short description (What? How?) 📖
We already fixed some MSVC warnings in #75 which was further optimized in #83. This got us so far that we could make an internal build with Vis. on Windows. But Filament contained many more public headers which we may not be using right now, but might need anytime in the future.

This PR fixes all W3 level MSVC warnings in the remaining public headers, regardless if it is used in Shapr or not.

## Upstreaming scope
[GFX-1891](https://shapr3d.atlassian.net/browse/GFX-1891)

### Design review 🎨
n/a

### Affected areas 🧭
n/a

### Special use-cases to test 🧷
n/a

### How did you test it? 🤔
Manual 💁‍♂️
I tested on VS2019 and V2022 with a .cpp file that includes all public headers of Filament (except for Apple/Linux only). You can test it too with https://github.com/shapr3d/shapr3d/tree/palver123/GFX-1868_filament_warnings_test (just have to change the path of your Filament git repo in Libraries.cmake)

Automated 💻

### CODE REVIEW

File: libs/image/include/image/ColorTransform.h [hunk 1]
```diff
@@ -231,8 +230,7 @@ inline std::unique_ptr<uint8_t[]> fromLinearToRGB_10_11_11_REV(const LinearImage
     using namespace filament::math;
     size_t w = image.getWidth();
     size_t h = image.getHeight();
-    UTILS_UNUSED_IN_RELEASE size_t channels = image.getChannels();
-    assert(channels >= 3);
+    assert(image.getChannels() >= 3);
```
Comment (palver123) [inline libs/image/include/image/ColorTransform.h:233]: In `fromLinearToGrayscale` below, they already made the choice of moving the function call inside the `assert`. I think it's ok to have it like this, no need for a local variable. The debugger can see the number of channels inside `image` anyway, so debugging experience is still _stellar_.

**Review by marcisolti — APPROVED** [review]
The modifications are LGTM. I built the `ALL_BUILD` target of win32 Filament, which revealed the following warning:

```
C:\git\filament\libs\utils\include\utils/Allocator.h(132,77): warning C4244: '=': conversion from 'uintptr_t' to 'uint32_t', possible loss of data
```
You can check this by compiling e.g. `OpenGLProgram.cpp`. 

However, I tried including `utils\Allocator.h` to `PhysicallyBasedRenderer.cpp`, but it did not cause a warning. I also tried using the function by instantiating a `LinearAllocator` and calling `alloc`, but still nothing. 

Doing
```
mCur = uint32_t(uintptr_t(p) - uintptr_t(mBegin))
```
eliminates the warning.

File: libs/utils/include/utils/BinaryTreeArray.h [hunk 2]
```diff
@@ -53,7 +53,7 @@ class BinaryTreeArray {
 public:
     static size_t count(size_t height) noexcept { return  (1u << height) - 1; }
     static size_t left(size_t i, size_t height) noexcept { return i + 1; }
-    static size_t right(size_t i, size_t height) noexcept { return i + (1u << (height - 1)); }
+    static size_t right(size_t i, size_t height) noexcept { return i + (1ull << (height - 1)); }
```
[the remark this answers is not in the document]
Comment (palver123) [reply libs/utils/include/utils/BinaryTreeArray.h]: Right. With C++23 I could use `1uz`/`1UZ` but for now we have to settle with `size_t(1)`. d728c1c1be3b86803f2fe6e6188a7519c8f681bf

**Comment by palver123** [comment]
@marcisolti you are right I forgot that we disable a bunch of warnings in Shapr3D codebase, so I tested with the wrong compiler flags. I update `palver123/GFX-1868_filament_warnings_test` and fixed 2 more warnings in 5ff9ee92687e04f0e000267a4d4df022668fe181 and 3f00a9ea3125c5a934bb2ce74c5cafc91b6124cf

**Review by veghbalazs — APPROVED** [review]

**Pull request merged**